src/runtime.rs

```diff
@@ -66,21 +80,98 @@ where
         self.increment_revision();
     }
 
+    /// Indicates that a derived query has begun to execute; if this is the
+    /// first derived query on this thread, then acquires a read-lock on the
+    /// runtime to prevent us from moving to a new revision until that query
+    /// completes.
+    ///
+    /// (However, if other threads invoke `increment_revision`, then
+    /// the current revision may be considered cancelled, which can be
+    /// observed through `is_current_revision_canceled`.)
+    pub(crate) fn freeze_revision(&self) -> Option<RwLockReadGuard<'_, ()>> {
+        let local_state = self.local_state.borrow();
+        if local_state.query_stack.is_empty() {
+            Some(self.shared_state.query_lock.read())
+        } else {
+            None
+        }
+    }
+
+    #[inline]
+    pub(crate) fn id(&self) -> RuntimeId {
+        self.id
+    }
+
     /// Read current value of the revision counter.
+    #[inline]
     pub(crate) fn current_revision(&self) -> Revision {
         Revision {
             generation: self.shared_state.revision.load(Ordering::SeqCst) as u64,
         }
     }
 
+    /// Check if the current revision is canceled. If this method ever
+    /// returns true, the currently executing query is also marked as
+    /// having an *untracked read* -- this means that, in the next
+    /// revision, we will always recompute its value "as if" some
+    /// input had changed. This means that, if your revision is
+    /// canceled (which indicates that current query results will be
+    /// ignored) your query is free to shortcircuit and return
+    /// whatever it likes.
+    #[inline]
+    pub fn is_current_revision_canceled(&self) -> bool {
+        let pending_revision_increments = self
+            .shared_state
+            .pending_revision_increments
+            .load(Ordering::SeqCst);
+        if pending_revision_increments > 0 {
+            self.report_untracked_read();
```

Comment:
Hm, will this be "transitive"?

I am thinking about a scenario where we try to revalidate query A, which executes (for validation) query B. Query B is then checks `is_current_revision_canceled`, sees that it is indeed cancelled and returns a dummy value, *which happens to match its previous value*. We then set `verified_at` to the current revision for `A`, which is not entirely correct. Maybe it's OK though?

Reply:
I think that's OK, but let's talk through it. 

So cancellation occurs in revision R1. Query A will be marked as verified-at for the current revision R1 (as will Query B). Then we will increment the revision to go to R2. Now we fetch Query A again -- it will check its inputs. Query B (which has an untracked read) will see that it had an untracked read and report itself as out of date. Query B will thus re-execute. *If it produces the same (dummy) value again even without cancellation*, then Query A will indeed not re-execute: but that's ok, it would have produced the same value. If it produces something different, then Query A should re-execute as normal.

Reply:
I was thinking about a slightly different scenario, when B is canceled during verification (that is, when cancellation happens at R2).

Let's say we verified both A and B at revision R1. Then, revision was advanced to R2, and we started to verify A again. When we were re-evaluating B at R2, cancellation occurred, so B returned a dummy value which happened to match that of R1 revision. Then, we mark A as verified at R2.

Reply:
OK, seems fine, since we don't care about the value of A at R2 (we cancelled R2).

Reply:
The question is: what happens to A in *R3*?

Reply:
In that case, when we re-demand A, it will go check its inputs, and B will get re-executed, etc.

Reply:
Put another way, I think it is correct not to re-execute A in R2 in that scenario, because its inputs did not change. But we ought to be sure that we re-execute A in R3 (presuming the result of B changes this time).